torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/matrix_ops.py

```diff
@@ -304,3 +313,100 @@ def sharded_bmm(args, kwargs, pg):
     extra_check=sharded_bmm_check,
     customized_func=sharded_bmm,
 )
+
+
+def sharded_layer_norm_check(*args, **kwargs):
+    """
+    Perform extra checks for the ``nn.LayerNorm`` op.
+    Ensure the normalized shape is compatible with
+    the size of the sharded tensor.
+
+    Args: same as ``torch.nn.LayerNorm``.
+
+    Return: None
+    """
+    st = args[0]
+    normalized_shape = args[1]
+    if st.dim() < len(normalized_shape):
+        raise ValueError(
+            "normalized_shape dim must not be greater than "
+            "the dim of the sharded tensor."
+        )
+    for idx in range(-1, -len(normalized_shape) - 1, -1):
+        if normalized_shape[idx] != st.size(idx):
+            raise ValueError(
+                f"Given normalized_shape=[{normalized_shape[idx]}], expected input with shape "
+                f"[*, {normalized_shape[idx]}], but got input of size {list(st.size())}."
+            )
+
+
+def sharded_layer_norm(args, kwargs, pg):
+    """
+    Handles ``__torch_function__`` dispatch for the ``torch.nn.LayerNorm`` op.
+    We gather all shards from local shards and perform a global normalization.
+    We then scatter the result back to each rank.
+
+    Args: same as ``torch.nn.LayerNorm``.
+
+    Return:
+        local_tensor (Tensor): New local tensor to build the sharded tensor.
+        sharding_spec (:class:`torch.distributed._shard.sharding_spec.ShardingSpec`):
+            sharding spec of the new sharded tensor.
+        new_st_size (torch.Size): Size of the new sharded tensor.
+    """
+    st = args[0]
+    normalized_shape = args[1]
+    sharding_dim = st.sharding_spec().dim  # type: ignore[attr-defined]
+    sharding_dim = sharding_dim if sharding_dim >= 0 else st.dim() + sharding_dim
+    local_tensor = st.local_tensor()
+    # If sharding dim is smaller than shape start, we just perform a local norm.
+    shape_start = st.dim() - len(normalized_shape)
+    if shape_start > sharding_dim:
+        args = (local_tensor, *args[1:])
+        local_tensor = torch.nn.functional.layer_norm(*args, **kwargs)
+        return local_tensor, st.sharding_spec(), st.size()
+
+    elementwise_affine = kwargs.get("elementwise_affine", False)
+    eps = kwargs.get("eps", 1e-05)
+
+    norm_dims = tuple(i for i in range(-1, -len(normalized_shape) - 1, -1))
+    local_size = math.prod(local_tensor.size()[shape_start:])  # type: ignore[attr-defined]
+    st_size = math.prod(st.size()[shape_start:])  # type: ignore[attr-defined]
+    local_mean = torch.mul(local_tensor.mean(norm_dims, keepdim=True), local_size)
+    global_mean = torch.div(all_reduce(local_mean), st_size)
+    local_variant_sq = torch.square(local_tensor - global_mean).sum(
+        norm_dims, keepdim=True
+    )
+    global_variant = torch.div(all_reduce(local_variant_sq), st_size)
```

Comment:
Do we need this additional allreduce? Shouldn't the global mean be sufficient to compute the global variance?

Reply:
I believe so, the way we calculate global variance still needs information from all shards. Unless there is some fancy math which can help us avoid the extra communication. Correct me if I am wrong.